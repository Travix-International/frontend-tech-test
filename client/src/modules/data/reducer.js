import { get, reduceStatePath } from '../lib';

export default (actions, dataStatePath) => {
  const {
    FETCH_PAGE,
    FETCH_PAGE_SUCCESS,
    FETCH_PAGE_FAILED,
    UPDATE_SUCCESS,
    DIRECT_UPDATE,
    DIRECT_UPDATE_SUCCESS,
    CANCEL_UPDATE,
    NEW,
    NEW_SUCCESS,
    NEW_CANCEL,
    REFRESH_PAGE,
    UPDATE,
    REQUEST_DELETE,
    CANCEL_DELETE,
    DELETE_SUCCESS,
    CONFIRM_DELETE,
    EDIT,
    FIELD_CHANGE,
    CHANGE_PAGE,
    SAVE_NEW,
    UPDATE_FAILED,
    DELETE_FAILED,
    NEW_FAILED,
    NEW_FIELD_CHANGE,
  } = actions;
  const defaultDataState = {
    pages: {},
    data: {},
    new: {},
  };
  const itemsReducer = action => state => {
    if (action.type === FETCH_PAGE) {
      const { page } = action.payload;
      return {
        ...state,
        pages: {
          ...state.pages,
          [page]: {
            ...state.pages[page],
            loading: get(
              state.pages,
              [page, 'loading'],
              true
            ),
            error: false,
          },
        },
      };
    }
    if (action.type === DELETE_SUCCESS) {
      return { ...state, pages: {}, data: {} };
    }
    if (action.type === FETCH_PAGE_SUCCESS) {
      const {
        result: {
          data,
          page: { current, total },
        },
      } = action.payload;
      return {
        ...state,
        pages: {
          ...state.pages,
          total,
          [current]: {
            error: false,
            loading: false,
            ids: data.map(item => item.id),
          },
        },
        data: data.reduce(
          (result, item) => ({
            ...result,
            [item.id]: item,
          }),
          state.data
        ),
      };
    }
    if (action.type === FETCH_PAGE_FAILED) {
      const { page, errorMessage } = action.payload;
      return {
        ...state,
        pages: {
          ...state.pages,
          [page]: {
            loading: false,
            error: true,
            errorMessage,
          },
        },
      };
    }
    if (
      action.type === UPDATE_SUCCESS ||
      action.type === DIRECT_UPDATE_SUCCESS
    ) {
      const { item } = action.payload;
      return {
        ...state,
        data: {
          ...state.data,
          [item.id]: item,
        },
      };
    }
    if (action.type === NEW) {
      return {
        ...state,
        new: {
          ...state.new,
          show: true,
          item: { title: '', description: '' },
        },
      };
    }
    if (action.type === NEW_FIELD_CHANGE) {
      const { fieldName, value } = action.payload;
      return {
        ...state,
        new: {
          ...state.new,
          item: { ...state.new.item, [fieldName]: value },
        },
      };
    }
    if (
      action.type === NEW_SUCCESS ||
      action.type === NEW_CANCEL
    ) {
      return {
        ...state,
        new: { show: false },
      };
    }
    if (action.type === SAVE_NEW) {
      return {
        ...state,
        new: {
          ...state.new,
          working: true,
        },
      };
    }
    if (action.type === REFRESH_PAGE) {
      const { page } = action.payload;
      return {
        ...state,
        pages: Object.entries(state.pages).reduce(
          (result, [key, value]) => {
            if (Number(key) !== page) {
              result[key] = value;
            }
            return result;
          },
          {}
        ),
      };
    }
    if (action.type === NEW_FAILED) {
      const { errorMessage: error } = action.payload;
      return {
        ...state,
        new: {
          ...state.new,
          working: false,
          error,
        },
      };
    }
    return state;
  };
  const componentReducer = (state, action) => {
    if (action.type === EDIT) {
      const { statePath, item } = action.payload;
      return reduceStatePath(
        state,
        statePath,
        itemState => ({
          ...itemState,
          edit: {
            ...itemState.edit,
            [item.id]: {
              item,
              error: false,
              working: false,
            },
          },
        })
      );
    }
    if (action.type === UPDATE) {
      const {
        statePath,
        item: { id },
      } = action.payload;
      return reduceStatePath(
        state,
        statePath,
        itemState => ({
          ...itemState,
          edit: {
            ...itemState.edit,
            [id]: {
              ...state[id],
              error: false,
              working: true,
            },
          },
        })
      );
    }
    if (action.type === UPDATE_FAILED) {
      const { errorMessage: error } = action.payload;
      const {
        statePath,
        item: { id },
      } = action.payload.fromAction.payload;
      return reduceStatePath(
        state,
        statePath.concat(['edit', id]),
        itemState => ({
          ...itemState,
          error,
          working: false,
        })
      );
    }
    if (
      action.type === UPDATE_SUCCESS ||
      action.type === CANCEL_UPDATE
    ) {
      const {
        statePath,
        item: { id },
      } = action.payload;
      return reduceStatePath(
        state,
        statePath,
        itemState => ({
          ...itemState,
          edit: Object.entries(itemState.edit).reduce(
            (result, [key, value]) => {
              if (Number(key) !== id) {
                result[key] = value;
              }
              return result;
            },
            {}
          ),
        })
      );
    }
    if (action.type === DELETE_FAILED) {
      const { errorMessage: error } = action.payload;
      const {
        statePath,
      } = action.payload.fromAction.payload;
      return reduceStatePath(
        state,
        statePath.concat('remove'),
        itemState => ({
          ...itemState,
          error,
          working: false,
        })
      );
    }
    if (action.type === REQUEST_DELETE) {
      const { item, statePath } = action.payload;
      return reduceStatePath(
        state,
        statePath,
        itemState => ({
          ...itemState,
          remove: {
            [item.id]: true,
            item,
          },
        })
      );
    }
    if (
      action.type === CANCEL_DELETE ||
      action.type === DELETE_SUCCESS
    ) {
      const { statePath } = action.payload;
      return reduceStatePath(
        state,
        statePath,
        itemState => ({
          ...itemState,
          remove: {},
        })
      );
    }
    if (action.type === CONFIRM_DELETE) {
      const { statePath } = action.payload;
      return reduceStatePath(
        state,
        statePath,
        itemState => ({
          ...itemState,
          remove: {
            ...itemState.remove,
            working: true,
          },
        })
      );
    }
    if (action.type === FIELD_CHANGE) {
      const {
        statePath,
        change: { id, fieldName, value },
      } = action.payload;
      return reduceStatePath(
        state,
        statePath.concat(['edit', id, 'item']),
        itemState => ({
          ...itemState,
          [fieldName]: value,
        })
      );
    }
    if (action.type === DIRECT_UPDATE) {
      const { statePath, item } = action.payload;
      return reduceStatePath(
        state,
        statePath.concat(['directUpdate']),
        itemState => ({
          ...itemState,
          [item.id]: item,
        })
      );
    }
    if (action.type === DIRECT_UPDATE_SUCCESS) {
      const {
        statePath,
        item: { id },
      } = action.payload;
      return reduceStatePath(
        state,
        statePath.concat(['directUpdate']),
        itemState =>
          Object.entries(itemState).reduce(
            (result, [key, value]) => {
              if (Number(key) !== id) {
                result[key] = value;
              }
              return result;
            },
            {}
          )
      );
    }

    if (action.type === CHANGE_PAGE) {
      const { statePath, page } = action.payload;
      return reduceStatePath(
        state,
        statePath,
        itemState => ({
          ...itemState,
          current: page,
        })
      );
    }
    return state;
  };

  return {
    defaultDataState,
    reducer: (state, action) =>
      reduceStatePath(
        componentReducer(state, action),
        dataStatePath,
        itemsReducer(action)
      ),
  };
};
