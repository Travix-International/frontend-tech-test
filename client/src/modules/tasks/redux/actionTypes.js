import { PENDING, FULFILLED, REJECTED } from 'redux-promise-middleware';

export const FETCH_TASKS = 'travix/front-end-test/tasks/FETCH_TASKS';
export const FETCH_TASKS_SUCCESS = `${FETCH_TASKS}_${FULFILLED}`;
export const FETCH_TASKS_ERROR = `${FETCH_TASKS}_${REJECTED}`;
export const FETCH_TASKS_LOADING = `${FETCH_TASKS}_${PENDING}`;
