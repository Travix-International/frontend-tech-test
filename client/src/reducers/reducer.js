import { GET_TASKS_SUCCESS, FILTER_TASKS, SORT_TASKS_BY_DESCRIPTION, SORT_TASKS_BY_TITLE } from '../constants/action-types';
const initialState = {
    tasks:[],
    filteredTasks:[],
    filter:""
};
// {id:0, title:'Task 1', description: 'wake up and shower!'}
// (
//     async () => {
//         const rawResponse = await fetch('/tasks', {
//         method: 'GET'
//         });
//         const content = await rawResponse.json();

//         console.log(content);
//     }
// )();

const tasks = (state = initialState, action) => {
    switch (action.type) {
        case 'CREATE_TASK':
            console.log(state);
            action.payload.id = state.tasks.length;
            return Object.assign({}, state, {
                tasks: state.tasks.concat(action.payload)
            });
        case 'EDIT_TASK':
            console.log(action);
            let newState = {...state};
            let editedTask = {...action.payload};
            console.log(editedTask);
            newState.tasks.splice(action.payload.id, 1, action.payload);
            console.log(newState);
            return newState;
        case 'DELETE_TASK':
            return Object.assign({}, state, {
                tasks: state.tasks.filter((task)=>task.id!==parseInt(action.payload.id, 10))
            });
        case GET_TASKS_SUCCESS:
            console.log(action);
            return Object.assign({}, state, {
                tasks: [...action.payload.tasks],
                filteredTasks: [...action.payload.tasks]
            });
        case FILTER_TASKS:
            console.log(action);
            let tasks = [...state.tasks];
            if (action.payload.length > 0) {
                let myRe = new RegExp(action.payload,"gi");
                let filteredTasks = tasks.filter((task)=>{
                    return (task.title.search(myRe) !==-1 || task.description.search(myRe) !==-1 ); 
                });
                return Object.assign({}, state, {
                    filteredTasks: filteredTasks,
                    filter: action.payload
                });
            } else {
                return Object.assign({}, state, {
                    filteredTasks: [...state.tasks],
                    filter: action.payload
                });
            }
        case SORT_TASKS_BY_DESCRIPTION:
            let allTasks = [...state.tasks];
            allTasks.sort((a, b)=>{
                if (a.description < b.description) {
                    return -1;
                } else if (a.description > b.description) {
                    return 1;
                } else {
                    return 0;
                }
            })
            return Object.assign({}, state, {
                tasks: allTasks
            });
        case SORT_TASKS_BY_TITLE:
            let preSortedTasks = [...state.tasks];
            preSortedTasks.sort((a, b)=>{
                if (a.title < b.title) {
                    return -1;
                } else if (a.title > b.title) {
                    return 1;
                } else {
                    return 0;
                }
            })
            return Object.assign({}, state, {
                tasks: preSortedTasks
            });
        default:
            return state
    }
}

export default tasks;