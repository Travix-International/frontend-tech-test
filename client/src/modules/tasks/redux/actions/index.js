import { createAction } from 'redux-actions';
import { FETCH_TASKS } from '../actionTypes';
import Tasks from '../../../../services/Tasks';

const tasksService = new Tasks();

export const fetchTasks = createAction(FETCH_TASKS, () => tasksService.fetchTasks());
