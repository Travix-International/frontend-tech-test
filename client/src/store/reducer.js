import * as actionTypes from './actionTypes';

const initialState = {
  tasks: [],
  selectedTask: null,
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case actionTypes.ALL_TASKS: {
      return {
        ...state,
        tasks: action.tasks.map(task => task),
      };
    }

    case actionTypes.TASK_ADDED: {
      return {
        ...state,
        tasks: [
          ...state.tasks,
          action.task,
        ],
      };
    }

    case actionTypes.TASK_UPDATED: {
      const tasks = state.tasks.map((task) => {
        if (task.id !== action.task.id) {
          return task;
        }
        return { ...action.task };
      });
      return {
        ...state,
        tasks,
      };
    }

    case actionTypes.TASK_DELETED: {
      return {
        ...state,
        tasks: [...state.tasks.filter(task => task.id !== action.id)],
      };
    }

    case actionTypes.SET_SELECT_TASK: {
      return {
        ...state,
        selectedTask: action.selectedTask,
      };
    }

    default:
      return state;
  }
};

export default reducer;
