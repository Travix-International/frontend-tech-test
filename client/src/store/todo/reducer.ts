import { Reducer } from 'redux';
import TodoState, { initialTodoState } from './state';
import { TodoAction, TodoActionTypes } from './types';

const { FETCH_TODO, FETCH_TODO_FAIL, FETCH_TODO_SUCCESS, ADD_TODO, TOGGLE_TODO, EDIT_TODO, DELETE_TODO } = TodoActionTypes;

const todoReducer: Reducer<TodoState, TodoAction> = (state = initialTodoState, action) => {
    switch (action.type) {
        case FETCH_TODO:
            return { ...state, pending: true };

        case FETCH_TODO_FAIL:
            return { ...state, pending: false, error: action.payload.error };

        case FETCH_TODO_SUCCESS:
            return { ...state, pending: false, todos: action.payload.todos };

        case ADD_TODO:
            return { ...state, todos: [...state.todos, action.payload.todo] };

        case TOGGLE_TODO:
            return {
                ...state, todos: state.todos.map(item =>
                    item.id === action.payload.id ? { ...item, done: !item.done } : item
                ),
            };

        case EDIT_TODO:
            return {
                ...state, todos: state.todos.map(item =>
                    item.id === action.payload.id ? { ...item, title: action.payload.title, description: action.payload.description } : item
                ),
            };

        case DELETE_TODO:
            return {
                ...state, todos: state.todos.filter(item => item.id !== action.payload.id),
            };

        default: return state;
    }
};

export default todoReducer;