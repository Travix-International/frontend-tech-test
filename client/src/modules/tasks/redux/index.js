import * as actionTypes from './actionTypes';
import * as factories from './factories';
import * as selectors from './selectors';
import * as utils from './utils';

import { fetchTasks } from './actions';

import reducerModule from './reducers';

const reducer = {
  name: 'tasks', reducer: reducerModule,
};

export {
  fetchTasks,
  actionTypes,
  factories,
  selectors,
  reducer,
  utils,
};
