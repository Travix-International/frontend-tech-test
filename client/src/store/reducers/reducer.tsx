import * as Actions from "../actions/actions";
import { Reducer, AnyAction } from "redux";
import { Task } from "../../models/Task";

interface State {
  tasks: Task[];
  selectedTask: Task | undefined;
}

let initialState: State = {
  tasks: [],
  selectedTask: undefined
};

export const tasksReducer: Reducer<any> = (
  state = initialState,
  action: AnyAction
) => {
  switch (action.type) {
    case Actions.FETCH_TASKS: {
      return { ...state, tasks: [...action.payload] };
    }
    case Actions.ADD_TASK: {
      return { ...state, tasks: [action.payload, ...state.tasks] };
    }
    case Actions.DELETE_TASK: {
      const filtered = state.tasks.filter(
        (task: Task) => task.id !== action.payload
      );
      return { ...state, tasks: [...filtered] };
    }
    case Actions.UPDATE_TASK: {
      state.tasks.find((task: Task) => {
        if (task.id === action.payload.id) {
          task.title = action.payload.title;
          task.description = action.payload.description;
          task.category = action.payload.category;
        }
      });
      return { ...state, tasks: [...state.tasks] };
    }
    case Actions.CHANGE_STATUS: {
      state.tasks.find((task: Task) => {
        task.id === action.payload && (task.completed = !task.completed);
      });
      return { ...state, tasks: [...state.tasks] };
    }
    case Actions.CHOOSE_TASK: {
      return { ...state, selectedTask: action.payload };
    }
    case Actions.FILTER_BY_CATEGORY: {
      const filtered = state.tasks.filter(
        (task: Task) => task.category.indexOf(action.payload) > -1
      );
      return { ...state, tasks: [...filtered] };
    }
    default:
      return state;
  }
};
