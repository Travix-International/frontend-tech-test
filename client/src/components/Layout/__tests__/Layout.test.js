import React from 'react'
import { createStore } from 'redux'
import { Provider } from 'react-redux'
import { render, cleanup } from 'react-testing-library'

import reducer from '../../../store/reducers/tasksReducer'
import { Layout } from '..'

describe('Layout should works properly', () => {
  function renderWithRedux(
    ui,
    { initialState, store = createStore(reducer, initialState) } = {},
  ) {
    return {
      ...render(<Provider store={store}>{ui}</Provider>),
      store,
    }
  }

  const fetchTasks = jest.fn()

  afterEach(cleanup)

  it('Should match snapshot', () => {
    const { container } = renderWithRedux(<Layout fetchTasks={fetchTasks} />)
    expect(container.firstChild).toMatchSnapshot('Layout_snapshot_1')
  })

  it('Should show a message when List is empty', () => {
    const { getByText } = renderWithRedux(
      <Layout fetchTasks={fetchTasks} loading={false} />,
    )
    getByText(
      'Your ToDo list is empty. Please add a new Task or just seat and enjoy the live',
    )
  })

  it('Should fetch tasks while mounting', () => {
    renderWithRedux(<Layout fetchTasks={fetchTasks} />)
    expect(fetchTasks).toHaveBeenCalled()
  })
})
