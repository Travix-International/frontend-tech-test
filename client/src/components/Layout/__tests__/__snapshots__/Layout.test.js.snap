// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout should works properly Should match snapshot: Layout_snapshot_1 1`] = `
<div
  class="Layout"
>
  <div
    class="Header"
  >
    <h1>
      Travix ToDo list
    </h1>
    <div
      class="Counter"
    >
      <p>
        Tasks To Do: 0
      </p>
    </div>
  </div>
  <div
    class="List"
  >
    <div>
      Your ToDo list is empty. Please add a new Task or just seat and enjoy the live
      <span
        aria-label="palm"
        role="img"
      >
        ðŸŒ´
      </span>
    </div>
  </div>
  <div
    class="Footer"
  >
    <div
      class="Inputs"
    >
      <input
        class="Input"
        name="title"
        placeholder="Your next task..."
        type="text"
        value=""
      />
      <textarea
        class="Input"
        name="desc"
        placeholder="Please write some description..."
        rows="5"
        type="text"
      />
    </div>
    <div
      class="Buttons"
    >
      <button
        class="Button_Add null null"
        disabled=""
        type="button"
      >
        add
      </button>
    </div>
  </div>
</div>
`;
